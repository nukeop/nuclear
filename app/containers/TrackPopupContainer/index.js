import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { ipcRenderer } from 'electron';

import * as DownloadsActions from '../../actions/downloads';
import * as FavoritesActions from '../../actions/favorites';
import * as PlayerActions from '../../actions/player';
import * as QueueActions from '../../actions/queue';
import * as ToastActions from '../../actions/toasts';

import ContextPopup from '../../components/ContextPopup';
import PopupButton from '../../components/ContextPopup/PopupButton';

const TrackPopupContainer = props => {
  const {
    trigger,
    track,
    artist,
    title,
    thumb,
    actions,
    musicSources,
    settings,

    withAddToQueue,
    withPlayNow,
    withAddToFavorites,
    withAddToDownloads
  } = props;
  
  return (
    <ContextPopup
      trigger={ trigger }
      artist={ artist }
      title={ title }
      thumb={ thumb }
    >
      {
        withAddToQueue &&
          <PopupButton
            onClick={() => {
              actions.addToQueue(
                musicSources,
                {
                  artist: props.artist,
                  name: props.title,
                  thumbnail: props.thumb
                });
            }}
            ariaLabel='Add track to queue'
            icon='plus'
            label='Add to queue'
          />
      }

      {
        withPlayNow &&
        <PopupButton
          onClick={ () => {
            actions.clearQueue();
            actions.addToQueue(
              musicSources,
              {
                artist: props.artist,
                name: props.title,
                thumbnail: props.thumb
              });
            actions.selectSong(0);
            actions.startPlayback();
          }}
          ariaLabel='Play this track now'
          icon='play'
          label='Play now'
        />
      }

      {
        withAddToFavorites &&
        <PopupButton
          onClick={ () => {
            actions.addFavoriteTrack(track);
            actions.info(
              'Favorite track added',
              `${artist} - ${title} has been added to favorites.`,
              <img src={ thumb } />,
              settings
            );
          }}
          ariaLabel='Add this track to favorites'
          icon='star'
          label='Add to favorites'
        />
      }

      {
        withAddToDownloads &&
        <PopupButton
          onClick={ () => {
            ipcRenderer.send('start-download', track);
            actions.addToDownloads(musicSources, track);
            actions.info(
              'Track added to downloads',
              `${artist} - ${title} has been added to downloads.`,
              <img src={ thumb } />,
              settings
            );
          }}
          ariaLabel='Download this track'
          icon='download'
          label='Download'
        />
      }
    
    </ContextPopup>
  );
};

TrackPopupContainer.propTypes = {
  trigger: PropTypes.node,
  track: PropTypes.object,
  artist: PropTypes.string,
  title: PropTypes.string,
  thumb: PropTypes.string,
  actions: PropTypes.shape({
    addToQueue: PropTypes.func,
    clearQueue: PropTypes.func,
    selectSong: PropTypes.func,
    startPlayback: PropTypes.func,
    addToDownloads: PropTypes.func
  }),
  musicSources: PropTypes.array,
  settings: PropTypes.object,

  withAddToQueue: PropTypes.bool,
  withPlayNow: PropTypes.bool,
  withAddToFavorites: PropTypes.bool,
  withAddToDownloads: PropTypes.bool
};

TrackPopupContainer.defaultProps = {
  trigger: null,
  track: {},
  artist: '',
  title: '',
  thumb: '',
  actions: {},
  musicSources: [],
  settings: {},
  
  withAddToQueue: true,
  withPlayNow: true,
  withAddToFavorites: true,
  withAddToDownloads: true
};

function mapStateToProps (state, { track }) {
  return {
    musicSources: track.local
      ? state.plugin.plugins.musicSources.filter(({ sourceName }) => {
        return sourceName === 'Local';
      })
      : state.plugin.plugins.musicSources,
    settings: state.settings
  };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators(
      Object.assign(
        {},
        DownloadsActions,
        FavoritesActions,
        QueueActions,
        PlayerActions,
        ToastActions
      ),
      dispatch
    )
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TrackPopupContainer);

