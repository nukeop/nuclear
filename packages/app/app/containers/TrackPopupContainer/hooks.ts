import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';
import { useTranslation } from 'react-i18next';

import { playlistsSelectors } from '../../selectors/playlists';
import { settingsSelector } from '../../selectors/settings';
import { pluginsSelectors } from '../../selectors/plugins';
import * as DownloadsActions from '../../actions/downloads';
import * as FavoritesActions from '../../actions/favorites';
import * as QueueActions from '../../actions/queue';
import * as ToastActions from '../../actions/toasts';
import * as PlaylistsActions from '../../actions/playlists';
import { getTrackArtist, getTrackItem } from '@nuclear/ui/lib';
import { safeAddUuid } from '../../actions/helpers';
import { PlaylistHelper } from '@nuclear/core';
import { TrackPopupStrings } from '@nuclear/ui/lib/components/TrackPopup';
export const useTrackPopupProps = (track, thumb) => {
  const playlists: Array<{name: string}> = useSelector(playlistsSelectors.playlists);
  const settings = useSelector(settingsSelector);
  const plugins: any = useSelector(pluginsSelectors.plugins);

  const streamProviders = track.local ? plugins.streamProviders.filter(({ sourceName }) => {
    return sourceName === 'Local';
  }) : plugins.streamProviders;

  const trackItem = getTrackItem(track);
  const dispatch = useDispatch();

  const onAddToQueue = useCallback(() => dispatch(
    QueueActions.addToQueue(trackItem)
  ), [trackItem, dispatch]);
  
  const onPlayNext = useCallback(() => dispatch(
    QueueActions.playNext(trackItem)
  ), [trackItem, dispatch]);

  const onPlayNow = useCallback(() => dispatch(
    QueueActions.playTrack(streamProviders, trackItem)
  ), [streamProviders, trackItem, dispatch]);

  const toastThumb = React.createElement('img', {src: thumb});
  const { t } = useTranslation('track-popup');
  const artist = getTrackArtist(track);

  const favoriteToastTitle = t('favorite-toast-title');
  const favoriteToastBody = t('favorite-toast-body', {
    artist,
    track: track?.name
  });
  const onAddToFavorites = useCallback(() => {
    dispatch(FavoritesActions.addFavoriteTrack(track));
    dispatch(ToastActions.info(
      favoriteToastTitle,
      favoriteToastBody,
      toastThumb,
      settings
    ));
  }, [dispatch, track, favoriteToastTitle, favoriteToastBody, toastThumb, settings]);

  const downloadToastTitle = t('download-toast-title');
  const downloadToastBody = t('download-toast-body', {
    artist,
    track: track?.name
  });
  const onAddToDownloads = useCallback(() => {
    const clonedTrack = {...safeAddUuid(trackItem)};
    dispatch(DownloadsActions.addToDownloads(streamProviders, clonedTrack));
    dispatch(ToastActions.info(
      downloadToastTitle,
      downloadToastBody,
      toastThumb,
      settings
    ));
  }, [trackItem, dispatch, streamProviders, downloadToastTitle, downloadToastBody, toastThumb, settings]);

  const playlistToastTitle = t('playlist-toast-title');
  const playlistToastBody = t('playlist-toast-body', {
    artist,
    track: track?.name
  });
  const onAddToPlaylist = useCallback((playlist) => {
    const clonedTrack = {...safeAddUuid(track)};
    if (clonedTrack.artist.name) {
      _.set(clonedTrack, 'artist', clonedTrack.artist.name);
    }
    
    playlist.tracks.push(PlaylistHelper.extractTrackData(track));
    dispatch(PlaylistsActions.updatePlaylist(playlist));
    dispatch(ToastActions.info(
      playlistToastTitle,
      `${playlistToastBody} ${playlist.name}.`,
      toastThumb,
      settings
    ));
  }, [track, dispatch, playlistToastTitle, playlistToastBody, toastThumb, settings]);

  const strings: TrackPopupStrings = {
    textAddToQueue: t('add-to-queue'),
    textPlayNow: t('play-now'),
    textPlayNext: t('play-next'),
    textAddToFavorites: t('add-to-favorite'),
    textAddToPlaylist: t('add-to-playlist'),
    textAddToDownloads: t('download')
  };

  return {
    playlists,
    strings,
    onAddToQueue,
    onPlayNext,
    onPlayNow,
    onAddToFavorites,
    onAddToPlaylist,
    onAddToDownloads
  };
};
