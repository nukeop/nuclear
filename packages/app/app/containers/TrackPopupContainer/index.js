import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { compose, withHandlers, withProps } from 'recompose';
import { TrackPopup, getTrackItem } from '@nuclear/ui';

import * as DownloadsActions from '../../actions/downloads';
import * as FavoritesActions from '../../actions/favorites';
import * as PlayerActions from '../../actions/player';
import * as QueueActions from '../../actions/queue';
import * as ToastActions from '../../actions/toasts';
import * as PlaylistsActions from '../../actions/playlists';
import { safeAddUuid } from '../../actions/helpers';
import { addTrackToPlaylist } from '../../components/PlayQueue/QueueMenu/QueueMenuMore';
import { getTrackArtist } from '@nuclear/ui/lib';

function mapStateToProps (state, { track }) {
  return {
    streamProviders: track.local
      ? state.plugin.plugins.streamProviders.filter(({ sourceName }) => {
        return sourceName === 'Local';
      })
      : state.plugin.plugins.streamProviders,
    settings: state.settings,
    playlists: state.playlists.playlists
  };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators(
      {
        ...DownloadsActions,
        ...FavoritesActions,
        ...QueueActions,
        ...PlayerActions,
        ...ToastActions,
        ...PlaylistsActions
      },
      dispatch
    )
  };
}

const TrackPopupContainer = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  withProps(({ track }) => ({
    trackItem: getTrackItem(track)
  })),
  withHandlers({
    onAddToQueue: ({ actions, trackItem }) => () => actions.addToQueue(trackItem),
    onPlayNow: ({ actions, streamProviders, trackItem }) => () => actions.playTrack(streamProviders, trackItem),
    onAddToFavorites: ({ actions, track, thumb, settings }) => () => {
      actions.addFavoriteTrack(track);
      actions.info(
        'Favorite track added',
        `${getTrackArtist(track)} - ${track.name} has been added to favorites.`,
        <img src={thumb} />,
        settings
      );
    },
    onAddToDownloads: ({ actions, streamProviders, trackItem, thumb, settings}) => () => {
      const clonedTrack = {...safeAddUuid(trackItem)};
      actions.addToDownloads(streamProviders, clonedTrack);
      actions.info(
        'Track added to downloads',
        `${getTrackArtist(clonedTrack)} - ${clonedTrack.name} has been added to downloads.`,
        <img src={thumb} />,
        settings
      );
    },
    onAddToPlaylist: ({ actions, settings, track, thumb}) => (playlist) => {
      const clonedTrack = {...safeAddUuid(track)};
      addTrackToPlaylist(actions.updatePlaylist, playlist, clonedTrack);
      actions.info(
        'Track added to playlist',
        `${getTrackArtist(clonedTrack)} - ${clonedTrack.name} has been added to playlist ${playlist.name}.`,
        <img src={thumb} />,
        settings
      );
    }
  })
)(TrackPopup);

TrackPopupContainer.propTypes = {
  trigger: PropTypes.node,
  track: PropTypes.object,
  artist: PropTypes.string,
  title: PropTypes.string,
  thumb: PropTypes.string,
  playlists: PropTypes.array,
  actions: PropTypes.shape({
    addToQueue: PropTypes.func,
    clearQueue: PropTypes.func,
    selectSong: PropTypes.func,
    startPlayback: PropTypes.func,
    updatePlaylist: PropTypes.func,
    addToDownloads: PropTypes.func
  }),
  streamProviders: PropTypes.array,
  settings: PropTypes.object,

  withAddToQueue: PropTypes.bool,
  withPlayNow: PropTypes.bool,
  withAddToFavorites: PropTypes.bool,
  withAddToPlaylist: PropTypes.bool,
  withAddToDownloads: PropTypes.bool
};

TrackPopupContainer.defaultProps = {
  trigger: null,
  track: {},
  artist: '',
  title: '',
  thumb: null,
  playlists: [],
  actions: {},
  streamProviders: [],
  settings: {},

  withAddToQueue: true,
  withPlayNow: true,
  withAddToFavorites: true,
  withAddToPlaylist: true,
  withAddToDownloads: true
};

export default TrackPopupContainer;
