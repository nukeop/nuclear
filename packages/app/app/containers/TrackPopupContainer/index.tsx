import React from 'react';

import { TrackPopup } from '@nuclear/ui';
import { ContextPopupProps } from '@nuclear/ui/lib/types';
import { useTrackPopupProps } from './hooks';
import { Track } from '@nuclear/core';

export type TrackPopupContainerProps = {
  track: Track;
  trigger: ContextPopupProps['trigger'];
  artist?: ContextPopupProps['artist'];
  title: ContextPopupProps['title'];
  thumb?: ContextPopupProps['thumb'];

  withAddToQueue?: boolean;
  withPlayNext?: boolean;
  withPlayNow?: boolean;
  withAddToFavorites?: boolean;
  withAddToPlaylist?: boolean;
  withAddToDownloads?: boolean;
};

const TrackPopupContainer: React.FC<TrackPopupContainerProps> = ({
  track=null,
  trigger=null,
  artist='',
  title='',
  thumb=null,
  withAddToQueue=true,
  withPlayNext=true,
  withPlayNow=true,
  withAddToFavorites=true,
  withAddToPlaylist=true,
  withAddToDownloads=true
}) => {
  const props = useTrackPopupProps(track, thumb);

  return <TrackPopup
    trigger={trigger}
    artist={artist}
    title={title}
    thumb={thumb}

    withAddToQueue={withAddToQueue}
    withPlayNext={withPlayNext}
    withPlayNow={withPlayNow}
    withAddToFavorites={withAddToFavorites}
    withAddToPlaylist={withAddToPlaylist}
    withAddToDownloads={withAddToDownloads}

    {...props}
  />;
};

export default TrackPopupContainer;
