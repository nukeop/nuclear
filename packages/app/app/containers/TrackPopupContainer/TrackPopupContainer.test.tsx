import React from 'react';
import userEvent from '@testing-library/user-event';
import TrackPopupContainer from './index';
import { buildStoreState } from '../../../test/storeBuilders';
import { AnyProps, mountComponent, setupI18Next } from '../../../test/testUtils';
import { RenderResult } from '@testing-library/react';

const TRACK_TITLE = 'Track title';
const TEST_IDS = {
  TRIGGER_BUTTON: 'trigger-button',
  ADD_TO_QUEUE_BUTTON: 'track-popup-add-queue',
  PLAY_NEXT_BUTTON: 'track-popup-play-next',
  PLAY_NOW_BUTTON: 'track-popup-play-now',
  ADD_TO_FAVORITES_BUTTON: 'track-popup-add-favorites',
  ADD_TO_PLAYLIST_BUTTON: 'track-popup-add-playlist',
  CREATE_PLAYIST_BUTTON: 'track-popup-create-playlist',
  DOWNLOAD_BUTTON: 'track-popup-download',
  CREATE_PLAYLIST_DIALOG_INPUT: 'track-popup-create-playlist-dialog-input',
  CREATE_PLAYLIST_DIALOG_ACCEPT: 'track-popup-create-playlist-dialog-accept',
  CREATE_PLAYLIST_DIALOG_CANCEL: 'track-popup-create-playlist-dialog-cancel'
};

const openTrackPopup = (component: RenderResult) => {
  userEvent.click(component.getByTestId(TEST_IDS.TRIGGER_BUTTON));
};

const openCreatePlaylistInputDialog = (component: RenderResult) => {
  userEvent.click(component.getByTestId(TEST_IDS.ADD_TO_PLAYLIST_BUTTON));
  const createPlaylistComponent = component.getByTestId(TEST_IDS.CREATE_PLAYIST_BUTTON);
  expect(createPlaylistComponent).toBeVisible();
  userEvent.click(createPlaylistComponent);
  expect(component.getByTestId(TEST_IDS.CREATE_PLAYLIST_DIALOG_INPUT)).toBeVisible();
};

describe('Track Popup container', () => {
  beforeAll(() => {
    setupI18Next();
  });

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it('should render the Track Popup container', () => {
    const { component } = mountTrackPopupContainer();
    openTrackPopup(component);
    expect(component.getByText(TRACK_TITLE)).toBeVisible();
    const popupButtonIds = [
      TEST_IDS.ADD_TO_QUEUE_BUTTON,
      TEST_IDS.PLAY_NEXT_BUTTON,
      TEST_IDS.PLAY_NOW_BUTTON,
      TEST_IDS.ADD_TO_FAVORITES_BUTTON,
      TEST_IDS.ADD_TO_PLAYLIST_BUTTON,
      TEST_IDS.DOWNLOAD_BUTTON
    ];
    popupButtonIds.forEach((id) => {
      expect(component.getByTestId(id)).toBeVisible();
    });
    expect(document.body).toMatchSnapshot(); // we need to use document.body because the popup is rendered outside the component
  });

  describe('creating new playlist', () => {
    it('should open Input Dialog', () => {
      const { component } = mountTrackPopupContainer();
      openTrackPopup(component);
      openCreatePlaylistInputDialog(component);
      expect(document.body).toMatchSnapshot();
    });

    it('should create new playlist with selected track', () => {
      const { component, store } = mountTrackPopupContainer();
      openTrackPopup(component);
      openCreatePlaylistInputDialog(component);
      const input = component.getByTestId(TEST_IDS.CREATE_PLAYLIST_DIALOG_INPUT).firstElementChild;
      expect(input).toHaveValue(TRACK_TITLE);
      const playlistName = 'New playlist name';
      userEvent.clear(input);
      userEvent.type(input, playlistName);
      expect(input).toHaveValue(playlistName);
      userEvent.click(component.getByTestId(TEST_IDS.CREATE_PLAYLIST_DIALOG_ACCEPT));

      const state = store.getState();      
      expect(state.playlists.localPlaylists.data).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: playlistName,
            tracks: [expect.objectContaining({ name: TRACK_TITLE })]
          })]));
    });

    it('should not create playlist with an empty name', async () => {
      const { component, store } = mountTrackPopupContainer();
      openTrackPopup(component);
      openCreatePlaylistInputDialog(component);
      const input = component.getByTestId(TEST_IDS.CREATE_PLAYLIST_DIALOG_INPUT).firstElementChild;
      userEvent.clear(input);
      expect(input).toHaveValue('');
      userEvent.click(component.getByTestId(TEST_IDS.CREATE_PLAYLIST_DIALOG_ACCEPT));

      const state = store.getState();
      expect(state.playlists.localPlaylists.data).not.toEqual([
        expect.objectContaining({
          name: ''
        })
      ]);
    });

    it('should add the song to next of current song when clicking PlayNext', async () => {
      const { component, store } = mountTrackPopupContainer({
        queue: {
          currentSong: 1,
          queueItems: [{
            artist: 'test artist 1',
            name: 'test track 1',
            streams: [{
              duration: 300,
              title: 'test track 1',
              skipSegments: []
            }]
          }, {
            artist: 'test artist 2',
            name: 'test track 2',
            streams: [{
              duration: 300,
              title: 'test track 2',
              skipSegments: []
            }]
          }]
        },
        player: {
          seek: 10
        }
      });
      
      openTrackPopup(component);
      userEvent.click(component.getByTestId(TEST_IDS.PLAY_NEXT_BUTTON));

      const state = store.getState();
      expect(state.player.seek).toBe(10);
      expect(state.queue.currentSong).toBe(1); 
      expect(state.queue.queueItems[state.queue.currentSong + 1]).toMatchObject({
        artist: 'Artist',
        name: TRACK_TITLE
      });
    });

  });


  const mountTrackPopupContainer = (initialStore?: AnyProps) => {
    return mountComponent(
      <TrackPopupContainer  
        title={TRACK_TITLE}
        track={{
          artist: {
            name: 'Artist'
          },
          title: TRACK_TITLE
        }}
        trigger={<button data-testid={TEST_IDS.TRIGGER_BUTTON}>Trigger</button>}
      />,
      ['/'],
      initialStore,
      buildStoreState()
        .build()
    );
  };
});
