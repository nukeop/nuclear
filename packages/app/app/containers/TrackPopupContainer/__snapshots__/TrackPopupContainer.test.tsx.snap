// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track Popup container creating new playlist should open Input Dialog 1`] = `
<body
  class="dimmable dimmed blurring"
>
  <div>
    <button
      data-testid="trigger-button"
    >
      Trigger
    </button>
  </div>
  <div
    style="display: flex; position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="ui bottom center popup transition visible context_popup"
      style="position: initial;"
    >
      <div
        class="popup_header"
      >
        <div
          class="popup_info"
        >
          <div
            class="popup_title"
          >
            Track title
          </div>
          
        </div>
      </div>
      <hr />
      <div
        class="popup_buttons"
      >
        <a
          aria-label="Add track to queue"
          class="popup_button"
          data-testid="track-popup-add-queue"
          href="#"
        >
          <i
            aria-hidden="true"
            class="plus icon"
          />
           
          Add to queue
        </a>
        <a
          aria-label="Add track to play next"
          class="popup_button"
          data-testid="track-popup-play-next"
          href="#"
        >
          <i
            aria-hidden="true"
            class="play icon"
          />
           
          Play next
        </a>
        <a
          aria-label="Play this track now"
          class="popup_button"
          data-testid="track-popup-play-now"
          href="#"
        >
          <i
            aria-hidden="true"
            class="play icon"
          />
           
          Play now
        </a>
        <a
          aria-label="Add this track to favorites"
          class="popup_button"
          data-testid="track-popup-add-favorites"
          href="#"
        >
          <i
            aria-hidden="true"
            class="heart icon"
          />
           
          Add to favorites
        </a>
        <div
          aria-expanded="false"
          class="ui dropdown nuclear dropdown popup_button left"
          data-testid="track-popup-add-playlist"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Add to playlist
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              class="item"
              data-testid="track-popup-create-playlist"
              role="option"
            >
              <i
                aria-hidden="true"
                class="plus icon"
              />
              Create new playlist
            </div>
          </div>
        </div>
        <a
          aria-label="Download this track"
          class="popup_button"
          data-testid="track-popup-download"
          href="#"
        >
          <i
            aria-hidden="true"
            class="download icon"
          />
           
          Download
        </a>
      </div>
    </div>
  </div>
  <div
    class="ui page modals dimmer transition visible active"
    style="display: flex;"
  >
    <div
      class="ui basic modal transition visible active"
    >
      <i
        aria-hidden="true"
        class="close icon"
      />
      <div
        class="content"
      >
        <h4>
          Input playlist name:
        </h4>
        <div
          class="ui fluid inverted input"
          data-testid="track-popup-create-playlist-dialog-input"
        >
          <input
            placeholder="Playlist name..."
            type="text"
            value="Track title"
          />
        </div>
      </div>
      <div
        class="actions"
      >
        <button
          class="ui red basic inverted button nuclear button"
          data-testid="track-popup-create-playlist-dialog-cancel"
        >
          Cancel
        </button>
        <button
          class="ui green button nuclear button"
          data-testid="track-popup-create-playlist-dialog-accept"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Track Popup container should render the Track Popup container 1`] = `
<body>
  <div>
    <button
      data-testid="trigger-button"
    >
      Trigger
    </button>
  </div>
  <div
    style="display: flex; position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="ui top left popup transition visible context_popup"
      style="position: initial;"
    >
      <div
        class="popup_header"
      >
        <div
          class="popup_info"
        >
          <div
            class="popup_title"
          >
            Track title
          </div>
          
        </div>
      </div>
      <hr />
      <div
        class="popup_buttons"
      >
        <a
          aria-label="Add track to queue"
          class="popup_button"
          data-testid="track-popup-add-queue"
          href="#"
        >
          <i
            aria-hidden="true"
            class="plus icon"
          />
           
          Add to queue
        </a>
        <a
          aria-label="Add track to play next"
          class="popup_button"
          data-testid="track-popup-play-next"
          href="#"
        >
          <i
            aria-hidden="true"
            class="play icon"
          />
           
          Play next
        </a>
        <a
          aria-label="Play this track now"
          class="popup_button"
          data-testid="track-popup-play-now"
          href="#"
        >
          <i
            aria-hidden="true"
            class="play icon"
          />
           
          Play now
        </a>
        <a
          aria-label="Add this track to favorites"
          class="popup_button"
          data-testid="track-popup-add-favorites"
          href="#"
        >
          <i
            aria-hidden="true"
            class="heart icon"
          />
           
          Add to favorites
        </a>
        <div
          aria-expanded="false"
          class="ui dropdown nuclear dropdown popup_button left"
          data-testid="track-popup-add-playlist"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="divider text"
            role="alert"
          >
            Add to playlist
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              class="item"
              data-testid="track-popup-create-playlist"
              role="option"
            >
              <i
                aria-hidden="true"
                class="plus icon"
              />
              Create new playlist
            </div>
          </div>
        </div>
        <a
          aria-label="Download this track"
          class="popup_button"
          data-testid="track-popup-download"
          href="#"
        >
          <i
            aria-hidden="true"
            class="download icon"
          />
           
          Download
        </a>
      </div>
    </div>
  </div>
</body>
`;
